src/express/controllers/api/action.ts

```diff
@@ -0,0 +1,105 @@
+import Review, {findCodeReviewRecord} from '../../../service/Review';
+import type {ChatPostMessageArguments} from '@slack/web-api';
+import type {RequestHandler} from 'express';
+import {formatApiCodeReviewRecord} from './reviews';
+import {postSlackMessage} from '../../../bolt/utils';
+
+export const actionController: RequestHandler = (req, res) => {
+    const action = req.params.action;
+    const value = req.params.value;
+
+    const {sid: slackUserId, fn: displayName} = req.user as {sid: string; fn: string};
+
+    if (!slackUserId) {
+        res.json({
+            error: 'Not authorized',
+        });
+        return;
+    }
+
+    if (['approve', 'claim', 'close', 'request-change', 'delete-record'].includes(action)) {
+        void findCodeReviewRecord({id: parseInt(value)}).then((codeReview) => {
+            if (!codeReview) {
+                res.json({error: 'Cannot find code review.'});
+                return;
+            }
+            switch (action) {
+                case 'approve':
+                    void Review.approve(codeReview, slackUserId).then((result) => {
+                        if (!result.codeReview) {
+                            res.json({error: result.message});
+                            return;
+                        }
+                        res.json({
+                            data: formatApiCodeReviewRecord(result.codeReview),
+                        });
+                        if (['inprogress', 'pending', 'ready'].includes(codeReview.status)) {
+                            void postSlackMessage(result.slackNotifyMessage as ChatPostMessageArguments);
+                        }
```

Comment:
Not a blocker, but it would be nice to DRY this up maybe with some helper functions

Reply:
Absolutely!